# Visual Data Analysis of Fraudulent Transactions

Your CFO has also requested detailed trends data on specific card holders. Use the starter notebook to query your database and generate visualizations that supply the requested information as follows, then add your visualizations and observations to your markdown report.

In [63]:
# Initial imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import calendar
import plotly.express as px
import hvplot.pandas
from sqlalchemy import create_engine

from dotenv import load_dotenv
import os

In [64]:
load_dotenv()

password = os.getenv("SQL_KEY")

# Create a connection to the database
engine = create_engine(f"postgresql://postgres:{password}@localhost:5432/fraud_detection")


## Data Analysis Question 1

The two most important customers of the firm may have been hacked. Verify if there are any fraudulent transactions in their history. For privacy reasons, you only know that their cardholder IDs are 2 and 18.

* Using hvPlot, create a line plot representing the time series of transactions over the course of the year for each cardholder separately. 

* Next, to better compare their patterns, create a single line plot that containins both card holders' trend data.  

* What difference do you observe between the consumption patterns? Does the difference suggest a fraudulent transaction? Explain your rationale in the markdown report.

In [65]:
# loading data for card holder 2 and 18 from the database
# Write the query

query = '''
    SELECT *
    FROM credit_card as cc
        INNER JOIN 
        card_holder as ch
        ON cc.card_holder_id = ch.card_holder_id
        INNER JOIN 
        transactions as t
        ON cc.card = t.card
        INNER JOIN
        merchant as m
        ON m.merchant_id = t.merchant_id
        INNER JOIN
        merchant_category as mc
        ON mc.merchant_category_id = m.merchant_category_id
    '''

# Create a DataFrame from the query result. HINT: Use pd.read_sql(query, engine)
df = pd.read_sql(query, engine)

_, i = np.unique(df.columns, return_index=True)
df = df.iloc[:, i]
df.head()

card  card_holder_id   card_holder_name category_name  \
0  3561954487988605              13        John Martin    food truck   
1  5135837688671496              13        John Martin           bar   
2   213193946980303              10  Matthew Gutierrez    food truck   
3  4263694062533017               4     Danielle Green           pub   
4     4498002758300              18      Malik Carlson    restaurant   

   merchant_category_id  merchant_id            merchant_name  \
0                     5           69            Dominguez PLC   
1                     3           85            Patton-Rivera   
2                     5           82               Day-Murray   
3                     4            5           Miller-Blevins   
4                     1           64  Cline, Myers and Strong   

   transaction_amount    transaction_date  transaction_id  
0                   6 2018-01-01 21:35:10             222  
1                   4 2018-01-01 21:43:12            2045  
2                  10 2018-01-01 22:41:21             395  
3                  19 2018-01-01 23:13:30            3309  
4                   3 2018-01-01 23:15:10             567

In [68]:
# Plot for cardholder 2
df_2 = df[df['card_holder_id'] == 2].loc[:, 'transaction_amount']
df_2.hvplot(kind='line')

:Curve   [index]   (transaction_amount)

In [69]:
# Plot for cardholder 18
df_18 = df[df['card_holder_id'] == 18].loc[:, 'transaction_amount']
df_18.hvplot(kind='line')

:Curve   [index]   (transaction_amount)

In [76]:
# Combined plot for card holders 2 and 18
df1 = pd.concat([df_2.reset_index(), df_18.reset_index()],
                axis=1).dropna().drop('index', axis=1)
df1.columns = ['cardholder 2 transactions', 'cardholder 3 transactions']
df1.hvplot(kind='line')

:NdOverlay   [Variable]
   :Curve   [index]   (value)

### Conclusion

The biggest difference between cardholder 2 & 3 is that cardholder 2 never makes big purchases, this could mean that they are frugal or they are trying not to get a big purchase flagged. 

Cardholder 3 however usually spends about the same amount as cardholder 2 however they have the occasional big purchase. Personally, I make big purchases every once and awhile when I have saved up some money, so cardholder 2 looks more realistic to me.

## Data Analysis Question 2

The CEO of the biggest customer of the firm suspects that someone has used her corporate credit card without authorization in the first quarter of 2018 to pay quite expensive restaurant bills. Again, for privacy reasons, you know only that the cardholder ID in question is 25.

* Using Plotly Express, create a box plot, representing the expenditure data from January 2018 to June 2018 for cardholder ID 25.

* Are there any outliers for cardholder ID 25? How many outliers are there per month?

* Do you notice any anomalies? Describe your observations and conclusions in your markdown report.

In [44]:
pd.date_range(start='2018-01-01', end='2018-07-01')

DatetimeIndex(['2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04',
               '2018-01-05', '2018-01-06', '2018-01-07', '2018-01-08',
               '2018-01-09', '2018-01-10',
               ...
               '2018-06-22', '2018-06-23', '2018-06-24', '2018-06-25',
               '2018-06-26', '2018-06-27', '2018-06-28', '2018-06-29',
               '2018-06-30', '2018-07-01'],
              dtype='datetime64[ns]', length=182, freq='D')

In [112]:
# loading data of daily transactions from jan to jun 2018 for card holder 25
# Write the query
query = '''
    SELECT *
    FROM credit_card as cc
        INNER JOIN 
        card_holder as ch
        ON cc.card_holder_id = 25
        INNER JOIN 
        transactions as t
        ON cc.card = t.card
        INNER JOIN
        merchant as m
        ON m.merchant_id = t.merchant_id
        INNER JOIN
        merchant_category as mc
        ON mc.merchant_category_id = m.merchant_category_id
    '''

# Create a DataFrame from the query result. HINT: Use pd.read_sql(query, engine)
df = pd.read_sql(query, engine, index_col='transaction_date').dropna().drop('card_holder_name', axis=1).sort_index()
df = df[(df.index >= '2018-01-01') & (df.index < '2018-07-01')]

_, i = np.unique(df.columns, return_index=True)
df = df.iloc[:, i]

df.head()

card  card_holder_id category_name  \
transaction_date                                                   
2018-01-02 02:06:21  4319653513507              25    food truck   
2018-01-02 02:06:21  4319653513507              25    food truck   
2018-01-02 02:06:21  4319653513507              25    food truck   
2018-01-02 02:06:21  4319653513507              25    food truck   
2018-01-02 02:06:21  4319653513507              25    food truck   

                     merchant_category_id  merchant_id     merchant_name  \
transaction_date                                                           
2018-01-02 02:06:21                     5           93  Rodriguez-Parker   
2018-01-02 02:06:21                     5           93  Rodriguez-Parker   
2018-01-02 02:06:21                     5           93  Rodriguez-Parker   
2018-01-02 02:06:21                     5           93  Rodriguez-Parker   
2018-01-02 02:06:21                     5           93  Rodriguez-Parker   

                     transaction_amount  transaction_id  
transaction_date                                         
2018-01-02 02:06:21                   1            2083  
2018-01-02 02:06:21                   1            2083  
2018-01-02 02:06:21                   1            2083  
2018-01-02 02:06:21                   1            2083  
2018-01-02 02:06:21                   1            2083

In [113]:
# loop to change the numeric month to month names
df.index = df.index.month_name()
df.head()

card  card_holder_id category_name  \
transaction_date                                                
January           4319653513507              25    food truck   
January           4319653513507              25    food truck   
January           4319653513507              25    food truck   
January           4319653513507              25    food truck   
January           4319653513507              25    food truck   

                  merchant_category_id  merchant_id     merchant_name  \
transaction_date                                                        
January                              5           93  Rodriguez-Parker   
January                              5           93  Rodriguez-Parker   
January                              5           93  Rodriguez-Parker   
January                              5           93  Rodriguez-Parker   
January                              5           93  Rodriguez-Parker   

                  transaction_amount  transaction_id  
transaction_date                                      
January                            1            2083  
January                            1            2083  
January                            1            2083  
January                            1            2083  
January                            1            2083

In [114]:
df.describe()

card_holder_id  merchant_category_id  merchant_id  transaction_amount  \
count          1700.0           1700.000000  1700.000000         1700.000000   
mean             25.0              3.117647    77.808824          135.514706   
std               0.0              1.450877    36.337335          372.447147   
min              25.0              1.000000    13.000000            1.000000   
25%              25.0              2.000000    47.750000            3.000000   
50%              25.0              3.000000    80.000000           10.000000   
75%              25.0              4.250000   103.250000           17.000000   
max              25.0              5.000000   149.000000         1813.000000   

       transaction_id  
count     1700.000000  
mean      1870.720588  
std        992.468428  
min        127.000000  
25%       1108.750000  
50%       1951.000000  
75%       2667.500000  
max       3437.000000

In [115]:
df['transaction_amount'].quantile(.87)

100.0

In [116]:
# Creating the six box plots using plotly express
fig = px.box(
    df.reset_index(),
    x='transaction_date',
    y='transaction_amount'
)

display(fig)

### Conclusion

At first glance we can draw the following conclusions:
- There are 1.5 outlier per month.
- There maximum payment in one transaction made in this period was 1,813 dollars with 86% of their transactions being 21 dollars or less.


We can also conclude from this data that there is significant evidence of anomalies in the transaction history of employee 25. At this point however, I wouldn't consider them anomolies. They spend 100 dollars or more 13% of time, this will add up significantly over time and raise the overall transaction average with it.